chain/src/main/java/org/ethereum/beacon/chain/SlotTicker.java

```diff
@@ -31,35 +30,41 @@
           .onBackpressureError()
           .name("SlotTicker.slot");
 
-  private static final long MILLIS_IN_SECOND = Duration.ofSeconds(1).toMillis();
+  private SlotNumber startSlot;
+
+  private Executor scheduler;
 
   public SlotTicker(SpecHelpers specHelpers, BeaconState state) {
     this.specHelpers = specHelpers;
-    this.accurateTime = specHelpers.accurateTimeMillis();
-    this.genesisTime = state.getGenesisTime();
+    this.state = state;
   }
 
   /** Execute to start {@link SlotNumber} propagation */
   @Override
   public void start() {
-    SlotNumber genesisSlot = specHelpers.getChainSpec().getGenesisSlot();
-    Time slotDuration = specHelpers.getChainSpec().getSlotDuration();
-    UInt64 slotDelta =
-        accurateTime
-            .get()
-            .dividedBy(UInt64.valueOf(MILLIS_IN_SECOND))
-            .minus(genesisTime)
-            .dividedBy(slotDuration);
-    UInt64 nextSlotStartTime = genesisTime.plus(slotDelta.increment()).times(slotDuration);
-    this.slot = genesisSlot.plus(slotDelta);
-    UInt64 delay = nextSlotStartTime.times(MILLIS_IN_SECOND).minus(accurateTime.get());
-    Flux.interval(Duration.ofSeconds(slotDuration.getIntValue()))
-        .delaySubscription(Duration.ofMillis(delay.longValue()))
-        .subscribe(
-            tick -> {
-              this.slot = slot.increment();
-              slotSink.onNext(new SlotNumber(slot));
+    this.scheduler =
+        Executors.newSingleThreadExecutor(
+            runnable -> {
+              Thread thread = new Thread(runnable, "slot-ticker");
+              thread.setDaemon(true);
+              return thread;
             });
+
+    SlotNumber nextSlot = specHelpers.get_current_slot(state).increment();
+    Time period = specHelpers.getChainSpec().getSlotDuration();
+
+    startImpl(nextSlot, period, Schedulers.fromExecutor(scheduler));
+  }
+
+  private void startImpl(SlotNumber startSlot, Time period, Scheduler scheduler) {
+    this.startSlot = startSlot;
+
+    Time startSlotTime = specHelpers.get_slot_start_time(state, startSlot);
+    long delayMillis = Math.max(0, startSlotTime.getMillis() - System.currentTimeMillis());
```

Comment:
please, use one time source in code
our application is time critical and will need to use one accurate source of time

Reply:
makes a lot of sense to me 👍